### PR TITLE
Implemented support for optional macro arguments

### PR DESCRIPTION
**Pull request by ln-zookeeper**
Some code review for this wouldn't hurt.

Example:

```
#define SET_LABEL X Y STRING

#arg TOOLTIP
"default "+{STRING}+" tooltip"
#endarg

#arg COLOR
255,255,255#endarg

    [label]
        x={X}
        y={Y}
        text={STRING}
        tooltip={TOOLTIP}
        color={COLOR}
    [/label]
#enddef

{SET_LABEL 25 23 "test label 1"}

{SET_LABEL 26 23 "test label 2" COLOR=0,0,255}

{SET_LABEL 27 23 "test label 3" (TOOLTIP= _ "this is a
multiline tooltip") COLOR=255,0,0}

{SET_LABEL 28 23 "test label 4" TOOLTIP=_"a single-line tooltip"}

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
Well, it looks okay, I guess. I don't see anything terrible in the code, though I feel like there's something missing with respect to error messages.

Have you verified that you can define inline macros with optional arguments and they'll still work? I mean macros intended to be used in the middle of an attribute, like {SPECIAL_NOTES}. For example:

```
#define SPECIAL_NOTES_PLAGUE
#arg TYPE
"Walking Corpse"#endarg
"Units killed by this unit rise as a {TYPE}."#enddef

description="This is a necromancer!" + {SPECIAL_NOTES_PLAGUE (Ghost)}
special_notes={SPECIAL_NOTES_PLAGUE (Ghost)} + {SPECIAL_NOTES_ARCANE}
```

(Not a great example since it uses a macro substitution within a string that should be translatable, which won't work correctly in practice; but it should get the idea across, and work as a test case.)

The key here is that, as I understand, the preprocessor is very precise about capturing whitespace in the macro definition (which is why the `#enddef` doesn't go on a separate line by itself as is the case with other macros). For this example to work (assuming my understanding is accurate), the `#arg` definition must contribute no whitespace to the macro content. It may already be the case, but it should be verified.

**Comment by gfgtdf** [comment]
the preproc_define contrustor takes all its parameters by const reference this means that all those objects will be copied later. This coudl be improved by taking `val`, `args` and `optargs` by rvalue reference and using std::move in the line that calls the constructor.

File: src/serialization/preprocessor.cpp [hunk 1]
```diff
@@ -1124,26 +1165,77 @@ bool preprocessor_data::get_chunk()
...
+						}
+					}
+				}
+
+				// If the macro definition has any optional arguments, insert their defaults
+				if (val.optional_arguments.size() > 0) {
+					for (const auto &arg : val.optional_arguments) {
+						if (defines->find(arg.first) == defines->end()) {
+							std::ostringstream res;
+							preprocessor_streambuf *buf = new preprocessor_streambuf(target_);
```
Comment (gfgtdf) [inline src/serialization/preprocessor.cpp]: Why do you use pointers here istead of just a local preprocessor_streambuf  object?

Comment (ln-zookeeper) [inline src/serialization/preprocessor.cpp]: Basically just to mirror the similar code found a bit further down. I can change buf to not be a pointer, but I figured it'd be better for readability if the declarations remain identical with the other places. Changing buffer seems to result in a crash.

**Comment by ln-zookeeper** [comment]
As for "inlined" arguments, yes those do work, there's no extra whitespace captured.

**Comment by Pentarctagon** [comment]
Is it possible to name the values being passed to the macro without making them optional?  For example:

> {SET_LABEL X=25 Y=23 STRING="test label 1"}

where X/Y/STRING wouldn't really have any meaning to Wesnoth, but could be useful for making it more clear a person reading the code what the macro is using the values for.

Or to put it another way, if I were to specify `X=25`, is it required to also specify an `#arg X`?

**Comment by ln-zookeeper** [comment]
No, that's not supported; you'd simply end up with "X=25" passed as the X argument and "Y=23" as the Y argument. That's partially a conscious decision, as that can't really be enabled without breaking compatibility with existing arguments which have "=" in them.

So, to be able to use X=25 you have to have an #arg X **and** you also can't have X in the normal argument list.

**Comment by gfgtdf** [comment]
So the 'optional' arugments _have_ to be passed with the NAME=VALUE, syntax while thenormal parameters cannot be passed with the NAME=VALUE syntax? I think that this is quite unexpected, although if allowing the NAME=VALUE syntax with nonoptional paremters woudl casue cmpablity issues its maybe the best we can offer right now.

**Comment by ln-zookeeper** [comment]
Yes, this doesn't affect how normal arguments work or how they can be used. It's certainly possible to add that kind of functionality later, but the compatibility issues would need to be solved somehow.

**Comment by Pentarctagon** [comment]
It doesn't seem like there would be a way to 100% solve any and all possible compatibility issues, since the values passed to macros can be pretty much arbitrary, but I think the best solution would be to use the required macro arguments to check if it's part of the value being passed or not.  So for example, if I had:

> #define MY_MACRO X Y

and I called it with:

> {MY_MACRO X=5 8}

then the preprocessor would know that the "X=" part shouldn't be used since the first argument defined for the macro is X and the value being passed starts with "X=".  But if I were to instead do:

> {MY_MACRO Z=5 8}

or

> {MY_MACRO 5 X=8}

or

> {MY_MACRO 5 MY=8}

then it would use the entire "Z=5" or "X=8" or "MY=8" value.

This would, I think, minimize any compatibility issues, since it would have to match both by the exact NAME and by position.

**Comment by CelticMinstrel** [comment]
It's definitely impossible to support keyword-based providing of required arguments within the current syntax without potentially breaking some uses. Anything using Wesnoth's normal conventions (all-uppercase, compared to the lowercase usually used for WML attributes) would not break, but quite a few technically valid but generally not used things would.

The only fully-compatible way I can think of to indicate a keyword-specified required argument would be to somehow use the sequence `{}` which, I assume, is currently not valid code. That would kinda look ugly, though.

The question then is whether we really care about a) perfect compatibility or b) keyword-specified required arguments. If we _do_ care about b), I would recommend tweaking this PR to reduce the potential compatibility issues of later allowing keyword-specified required arguments. Specifically, the current way of specifying optional arguments looks exactly like a valid WML key, and WML keys are somewhat common arguments to macros. A slightly different syntax would reduce the chance of conflict in the event of later allowing keyword-specified required arguments.

For example, instead of `NAME=VALUE`, you could use `NAME:=VALUE` or maybe `NAME->VALUE`. Neither of those look like a valid WML key, which would help for future compatibility in the event of extending the syntax to required arguments. (True, this is only really a compatibility issue if people use lower-case macro argument names, but who are we to tell them that they absolutely must never do that?)

**Comment by Pentarctagon** [comment]
> This completely defeats the point of allowing it in my opinion, since the main argument for it is freeing the call site from needing to know or care about the order of arguments.

The only reason I was interested in it was from a purely non-functional perspective - it would allow providing more information to the person reading the code without having to go look at what the macro was actually using the value for.  So if I saw some macro with `X=5 Y=5`, I could reasonably assume that is a hex coordinate, regardless of what the macro itself is named.  Or perhaps a more concrete example, if I had a macro called `CREATE_UNIT_CLUSTER` which I then called with:

`{CREATE_UNIT_CLUSTER 3 7 5 "Spearman"}`

it would be much more informative to instead write it as:

`{CREATE_UNIT_CLUSTER X=3 Y=7 NUMBER=5 "Spearman"}`

To me, required arguments should also have a required order, and while I can see how the ability to specify arguments in whatever order you want could be useful, actually making consistent use of that ability seems like it would end up being rather confusing.

> True, this is only really a compatibility issue if people use lower-case macro argument names, but who are we to tell them that they absolutely must never do that?

Personally I've always thought that the convention to have macro names and arguments be uppercase is so widely used that it might as well be enforced by the preprocessor - in fact I had assumed it _was_ actually a rule for a long time, until I looked through the WML of the original Orocia add-on and saw macros with lowercase names.  But this is neither here nor there, I suppose.

**Comment by ln-zookeeper** [comment]
I'm really not convinced that this should even be available for required arguments. While obviously `X=5` is more informative than `5`, it'd also be ambiguous whether it's a required or optional argument. If that form worked, I think actually using it might be rather foolish because it could likely just make things less clear, at least to a reader who needs to understand how the macro works, even if not to a casual reader who just wants to be able to tell what the code does.

Anyway, as far as accommodating possible future expansion goes, I think that can then be done by simply checking that the `NAME` in `NAME=VALUE` matches one of the macro's required arguments. So, `x,y=7,9` wouldn't be mistaken as the new syntax because the macro wouldn't have an argument called x,y, whereas the alternative form `LOC=x,y=7,9` (assuming the argument name is LOC) would work. Arguably it looks pretty ugly, but one can always add spacing `(LOC= x,y=7,9)` if they prefer. Neither `NAME->VALUE` nor `NAME:=VALUE` strike me as particularly pretty, either.

**Comment by ln-zookeeper** [comment]
I added a check for whether the passed optional argument actually matches the name of any of the optional arguments the macro has declared, and had it output a non-fatal warning if it doesn't.

It was also previously possible to (ab)use the syntax to temporarily (for the duration of the macro) overwrite the value of any other define that the macro was calling internally, which now won't happen.

**Comment by gfgtdf** [comment]
>```
> {SET_LABEL 27 23 "test label 3" (TOOLTIP= _ "this is a
> multiline tooltip") COLOR=255,0,0}
> ```


is  
```
{SET_LABEL 27 23 "test label 3" TOOLTIP=(_ "this is a
multiline tooltip") COLOR=255,0,0}
```
also supported?

**Comment by CelticMinstrel** [comment]
I would guess `TOOLTIP=_"this is a tooltip"` is also supported (and also with added line break). Note the lack of space around the _.

**Comment by ln-zookeeper** [comment]
No, `NAME=(VALUE)` won't work, if you need parentheses then it has to be `(NAME=VALUE)`. To change that would be quite a bit more difficult, because then at least some of the logic would have to be extended into the lower-level code which decides how arguments are delimited.

And yes, also `NAME=_"VALUE"` works.

**Comment by CelticMinstrel** [comment]
I always assumed that parentheses anywhere in the argument protect spaces... but I don't think I've ever actually tried it...

**Pull request merged**